spec/features/user_can_sign_up_spec.rb

```diff
@@ -0,0 +1,67 @@
+require "rails_helper"
+
+RSpec.feature "SignUp", type: :feature do
+  scenario "Can sign up and details are saved" do
+    visit '/'
+    within '.header' do
+      click_link "Sign Up"
+    end
+    expect(page).to have_button("Sign Up")
+    fill_in "Username", with: "Username"
+    fill_in "Email", with: "user@example.com"
+    fill_in "Password", with: "Password"
+    fill_in "user_password_confirmation", with: "Password"
+    click_button "Sign Up"
+    expect(current_path).to eq "/posts"
+
+    user = User.find_by(username: "Username")
+
+    expect(user.username).to eq "Username"
+    expect(user.email).to eq "user@example.com"
+    expect(user.password).not_to eq "Password"
+  end
+
+  scenario "password is hidden" do
+    visit '/'
+    within '.header' do
+      click_link "Sign Up"
+    end
+    fill_in "Username", with: "Username"
+    fill_in "Email", with: "user@example.com"
+    fill_in "Password", with: "hidden"
+    expect(page).to_not have_content("hidden")
+    fill_in "user_password_confirmation", with: "hidden"
+    expect(page).to_not have_content("hidden")
+  end
+
+  scenario "cannot sign up if username not unique" do
+    user = User.create(username: "username", email: "user1@example.com", password: "password")
+    visit '/'
+    within '.header' do
+      click_link "Sign Up"
+    end
+    fill_in "Username", with: "username" #duplicate
+    fill_in "Email", with: "user@example.com"
+    fill_in "Password", with: "hidden"
+    expect(page).to_not have_content("hidden")
+    fill_in "user_password_confirmation", with: "hidden"
+    click_button "Sign Up"
+    expect(page).to have_content("Sign up")
+  end
+
+  scenario "cannot sign up if username not unique" do
+    user = User.create(username: "username", email: "user1@example.com", password: "password")
+    visit '/'
+    within '.header' do
+      click_link "Sign Up"
+    end
+    fill_in "Username", with: "username2"
+    fill_in "Email", with: "user1@example.com" #duplicate
+    fill_in "Password", with: "hidden"
+    expect(page).to_not have_content("hidden")
+    fill_in "user_password_confirmation", with: "hidden"
+    click_button "Sign Up"
+    expect(page).to have_content("Sign up")
```

Comment:
For these 2 tests i wonder is this the most appropriate expect statement to confirm that the user hasn't been created. Could you maybe do a User.find with the duplicate email and then expect that the username is not equal to the one entered? Maybe a bit pedantic of me

Reply:
Yes I was wondering that but then I thought that for a feature test you shouldn't check the back end? It's all from a user perspective? So actually then perhaps I should have tried to log in and expect that not to work? 🤔